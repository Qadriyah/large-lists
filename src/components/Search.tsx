/* eslint-disable react-hooks/exhaustive-deps */
"use client";
import { debounce } from "lodash";
import { useRouter, useSearchParams } from "next/navigation";
import React, { useTransition } from "react";
import SearchIcon from "./SearchIcon";
import Spinner from "./Spinner";

const Search = () => {
  const searchParams = useSearchParams();
  const router = useRouter();
  const [isPending, startTransition] = useTransition();

  const handleChange = React.useCallback(
    debounce((event: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = event.target;
      const newSearchParams = new URLSearchParams(searchParams.toString());
      if (!value) {
        newSearchParams.delete("country");
      } else {
        newSearchParams.set("country", event.target.value);
      }
      startTransition(() => router.push(`?${newSearchParams.toString()}`));
    }, 500),
    []
  );

  console.log(isPending);

  return (
    <div className="flex">
      <div className="flex gap-2 h-8 px-2 items-center border border-gray-300 rounded-md">
        {isPending ? (
          <Spinner />
        ) : (
          <SearchIcon width="18px" height="18px" className="text-gray-500" />
        )}
        <input
          type="search"
          placeholder="search..."
          onChange={handleChange}
          defaultValue={searchParams.get("country") || ""}
          className="border-none outline-none"
        />
      </div>
    </div>
  );
};

export default Search;
