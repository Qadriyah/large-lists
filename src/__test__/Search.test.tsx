import Search from "@/components/Search";
import { render } from "@testing-library/react";

describe("Search", () => {
  test("renders the search field", () => {
    const { container } = render(<Search searchParam="country" />);
    expect(container).toMatchSnapshot();
  });
});
