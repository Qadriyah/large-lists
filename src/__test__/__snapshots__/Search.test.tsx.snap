// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders the search field 1`] = `
<div>
  <div
    class="flex"
  >
    <div
      class="flex gap-2 h-8 px-2 items-center border border-gray-300 rounded-md"
    >
      <svg
        class="text-gray-500"
        fill="none"
        height="18px"
        stroke="currentColor"
        stroke-width="3"
        viewBox="0 0 24 24"
        width="18px"
      >
        <path
          d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
        />
      </svg>
      <input
        class="border-none outline-none"
        placeholder="search by country ..."
        type="search"
        value=""
      />
    </div>
  </div>
</div>
`;
